### PR TITLE
feat: post process AZAddSecret edges from roles instead of principals

### PR DESCRIPTION
**Pull request by mistahj67**
## Description

Add AZAddSecret edges from AZRoles to AZApps instead of AZUsers / AZGroups / AZServicePrincipals

## Motivation and Context

Going from the principal entities down to the apps themselves is not as performant in post processing as from the roles. 

## How Has This Been Tested?

Added integration test with AZAddSecret harness, screenshot below

## Screenshots (if appropriate):
AZ Add Secret

## Types of changes

-   [ ] Chore (a change that does not modify the application functionality)
-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [ ] Documentation updates are needed, and have been made accordingly.
-   [x] I have added and/or updated tests to cover my changes.
-   [x] All new and existing tests passed.
-   [ ] My changes include a database migration.

### CODE REVIEW

File: cmd/api/src/analysis/azure/azure_integration_test.go [hunk 1]
```diff
@@ -648,6 +648,20 @@ func TestRoleEntityDetails(t *testing.T) {
...
+	testContext := integration.NewGraphTestContext(t, schema.DefaultGraphSchema())
+	testContext.ReadTransactionTestWithSetup(func(harness *integration.HarnessDetails) error {
+		harness.AZAddSecretHarness.Setup(testContext)
+		return nil
+	}, func(harness integration.HarnessDetails, tx graph.Transaction) {
+
+		postProcessingStats, err := azureanalysis.AppRoleAssignments(context.Background(), testContext.Graph.Database)
+		require.Nil(t, err)
+		require.NotNil(t, postProcessingStats.RelationshipsCreated[azure.AddSecret])
+		assert.Equal(t, 8, int(*postProcessingStats.RelationshipsCreated[azure.AddSecret]))
```
Comment (mistahj67) [inline cmd/api/src/analysis/azure/azure_integration_test.go:661]: This likely could use a tighter test verifying which edges are to which role but I wasn't entirely sure the best way to query for that.

File: packages/go/analysis/azure/role.go [hunk 2]
```diff
@@ -59,3 +66,237 @@ func PopulateRoleEntityDetailsCounts(tx graph.Transaction, node *graph.Node, det
...
+
+	if tenantRoles, err := TenantRoles(tx, tenant, roleTemplateIDs...); err != nil {
+		return nil, err
+	} else {
+		return roleMembers(tx, tenantRoles, azure.GrantSelf)
+	}
+}
+
+// RoleScopedTo returns the NodeSet of scoped to entities for a given role.
+func RoleScopedTo(tx graph.Transaction, role *graph.Node) (graph.NodeSet, error) {
```
Comment (mistahj67) [inline packages/go/analysis/azure/role.go:286]: This is new, the rest is just moved here to better match our file structure

File: packages/go/analysis/azure/tenant.go [hunk 3]
```diff
@@ -81,6 +82,62 @@ func TenantPrincipals(tx graph.Transaction, tenant *graph.Node) (graph.NodeSet,
...
+		conditions = append(conditions, query.In(query.EndProperty(azure.RoleTemplateID.String()), roleTemplateIDs))
+	}
+
+	return ops.FetchEndNodes(tx.Relationships().Filterf(func() graph.Criteria {
+		return query.And(conditions...)
+	}))
+}
+
+// TenantApplications returns the complete set of application nodes contained by the given Tenant node
+func TenantApplications(tx graph.Transaction, tenant *graph.Node) (graph.NodeSet, error) {
```
Comment (mistahj67) [inline packages/go/analysis/azure/tenant.go:127]: This is new, the rest is just moved here to better match our file structure

**Review by ddlees — APPROVED** [review]
LGTM; If you can beef up the integration test to verify the nodes and edges apart from the post-processing stats that would be great but I'm not all that worried about it if it's too much work. Better to get this in sooner to relieve pressure given how long it currently takes to process AZAddSecret.

**Pull request merged**